vms/platformvm/txs/executor/proposal_tx_executor.go

```diff
@@ -320,215 +322,303 @@ func (e *ProposalTxExecutor) RewardValidatorTx(tx *txs.RewardValidatorTx) error
-			if err != nil {
-				return fmt.Errorf("failed to create output: %w", err)
-			}
-			out, ok := outIntf.(verify.State)
-			if !ok {
-				return ErrInvalidState
-			}
+	// Provide the reward here
+	rewardToPayBack := validator.PotentialReward
+	if rewardToPayBack > 0 {
+		validationRewardsOwner := uValidatorTx.ValidationRewardsOwner()
+		outIntf, err := e.Fx.CreateOutput(rewardToPayBack, validationRewardsOwner)
+		if err != nil {
+			return fmt.Errorf("failed to create output: %w", err)
+		}
+		out, ok := outIntf.(verify.State)
+		if !ok {
+			return ErrInvalidState
+		}
 
-			utxo := &avax.UTXO{
-				UTXOID: avax.UTXOID{
-					TxID:        tx.TxID,
-					OutputIndex: uint32(len(outputs) + len(stake)),
-				},
-				Asset: stakeAsset,
-				Out:   out,
-			}
+		utxo := &avax.UTXO{
+			UTXOID: avax.UTXOID{
+				TxID:        txID,
+				OutputIndex: uint32(utxosOffset),
+			},
+			Asset: stakeAsset,
+			Out:   out,
+		}
+		e.OnCommitState.AddUTXO(utxo)
+		e.OnCommitState.AddRewardUTXO(txID, utxo)
 
-			e.OnCommitState.AddUTXO(utxo)
-			e.OnCommitState.AddRewardUTXO(tx.TxID, utxo)
+		utxosOffset++
+	}
 
-			offset++
-		}
+	// Provide the accrued delegatee rewards from successful delegations here.
+	delegateeRewardToPayBack, err := e.OnCommitState.GetDelegateeReward(
+		validator.SubnetID,
+		validator.NodeID,
+	)
+	if err != nil {
+		return fmt.Errorf("failed to fetch accrued delegatee rewards: %w", err)
+	}
 
-		// Provide the accrued delegatee rewards from successful delegations here.
-		delegateeReward, err := e.OnCommitState.GetDelegateeReward(
-			stakerToRemove.SubnetID,
-			stakerToRemove.NodeID,
-		)
+	if delegateeRewardToPayBack > 0 {
+		delegationRewardsOwner := uValidatorTx.DelegationRewardsOwner()
+		outIntf, err := e.Fx.CreateOutput(delegateeRewardToPayBack, delegationRewardsOwner)
 		if err != nil {
-			return fmt.Errorf("failed to fetch accrued delegatee rewards: %w", err)
+			return fmt.Errorf("failed to create output: %w", err)
+		}
+		out, ok := outIntf.(verify.State)
+		if !ok {
+			return ErrInvalidState
 		}
 
-		if delegateeReward > 0 {
-			delegationRewardsOwner := uStakerTx.DelegationRewardsOwner()
-			outIntf, err := e.Fx.CreateOutput(delegateeReward, delegationRewardsOwner)
-			if err != nil {
-				return fmt.Errorf("failed to create output: %w", err)
-			}
-			out, ok := outIntf.(verify.State)
-			if !ok {
-				return ErrInvalidState
-			}
+		onCommitUtxo := &avax.UTXO{
+			UTXOID: avax.UTXOID{
+				TxID:        txID,
+				OutputIndex: uint32(utxosOffset),
+			},
+			Asset: stakeAsset,
+			Out:   out,
+		}
+		e.OnCommitState.AddUTXO(onCommitUtxo)
+		e.OnCommitState.AddRewardUTXO(txID, onCommitUtxo)
+
+		// Note: There is no [offset] if the RewardValidatorTx is
+		// aborted, because the validator reward is not awarded.
+		onAbortUtxo := &avax.UTXO{
+			UTXOID: avax.UTXOID{
+				TxID:        txID,
+				OutputIndex: uint32(utxosOffset - 1),
```

Comment:
isn't this wrong if rewardToPayBack =< 0?

Reply:
I believe it's correct.
Current code uses `len(outputs) + len(stake) + offset` for onCommitUTXO and `len(outputs) + len(stake) ` for the abort utxo.
My refactoring uses `utxosOffset` for onCommintUtxo, so it must decrement the onAbortUtxo by 1

Reply:
but if `rewardToPayBack` is < 0 then `utxosOffset` will equal to `len(outputs) + len(stake)`, since it won't be incremented. then `utxsoOffset` - 1 == `len(outputs) + len(stake)` - 1 instead of  len(outputs) + len(stake)

Reply:
Reverted.